### DISCUSSION TITLE
Trouble with telemetry on ESP32C3

### DISCUSSION THREAD

**Discussion by GTorreil** [ZZ-Cat/CRSFforArduino#142] [Q&A]
Hello ! 

I'm playing with an ESP32C3 and this library. I'm using serial 1 for CRSF. The receiver part worked first-try and i've flown a plane several time with this setup as the receiving end -- it works flawlessly. However, I can't get any telemetry back. I do get the receiver's telemetry (`1RSS`, `2RSS`, and so on) but my calls to `telemetryWriteBattery` do not seem to send anything back. 

Checklist : 

- [x] Duplicate ? Not really a duplicate but this may be related to https://github.com/ZZ-Cat/CRSFforArduino/discussions/127 . Although I somehow doubt that since the Rx part works fine. BTW I can totally confirm Rx works on ESP32C3 🥳 
- [x] I did "discover" new sensors on the transmitter side (RM Pocket and RM TX16S - same results)
- [x] I tried calling `telemetryWriteBattery` from `loop()`, from `loop()` but with `millis()`-based timing to limit the rate to 10Hz, and from `onReceiveRcChannels`
- [x] I read the [Telemtry docs](https://github.com/ZZ-Cat/CRSFforArduino/wiki/API-Reference#telemetry)
- [x] Edit : I also upgraded Rx and Tx to latest ELRS versions 

[Repro](https://gist.github.com/GTorreil/c24ef5b04a68d763804c815da8e70dca) (or is it ? Maybe i'm just stupid)

Thanks for the feedback on this one !

**Comment by ZZ-Cat** [comment]
In your [gist here](https://gist.github.com/GTorreil/c24ef5b04a68d763804c815da8e70dca#file-main-cpp-L19-L20), calling `Serial1.begin()` is the likely culprit.  
`Serial1.begin()` is already internally called in `CRSFforArduino::begin()`, so calling `Serial1.begin()` in your sketch is unnecessary. It looks like you have effectively "confused" the firmware by initialising the port with 400K baud and pins 20 and 21 being your Tx and Rx pins, then in the line below it, `CRSFforArduino::begin()` is initialising `Serial1` again with 420K baud and pins 0 and 1 as your Tx and Rx pins.  
To verify this, you can go into [CFA_Config.hpp](https://github.com/ZZ-Cat/CRSFforArduino/blob/Main-Trunk/src/CFA_Config.hpp), set the `DEBUG_ENABLED` (I realise I accidentally deleted this flag a while ago in the config. :sweat_smile: So, go ahead and add it back in) and `CRSF_DEBUG_ENABLE_CONFIGURATION_DUMP` flags to `1`, build-and-flash your ESP32C3 with these flags enabled, and it will show the actual configuration in the Serial Monitor - the actual configuration may not be what you intend it to be.

All ESP32 targets support custom assignable Serial Tx and Rx pins.  
You can pass in your desired pins directly into CFA's [constructor](https://github.com/ZZ-Cat/CRSFforArduino/wiki/API-Reference#constructors), then you can [instantiate it](https://github.com/ZZ-Cat/CRSFforArduino/wiki/API-Reference#instantiation) with your custom pin settings.  
Doing it this way actually tells CFA what pins to use and how to configure them.

10 Hz telemetry updates should be fine. That's what I tested CFA on in the [Telemetry Example Sketch](https://github.com/ZZ-Cat/CRSFforArduino/blob/eeeef277720f41239797f5afc6694179233e0bd4/examples/telemetry/telemetry.ino#L147-L177).

Calling the `telemetryWrite*` functions within the context of the `onReceiveRcChannels()` event has no benefit, because telemetry is actually written out when `CRSFforArduino::update()` is called.
